### ISSUE TITLE
State of mtl

### ISSUE THREAD

**Issue by phadej** [haskell/core-libraries-committee#52]
> If CLC finds a Core Library in a neglected state, it can source and install new maintainers or resolve situation by other means.

 `mtl` is in neglected state. The weren't `transformers-0.6` compatible release yet, though [`transformers-0.6`](https://hackage.haskell.org/package/transformers-0.6.0.0) was released in June 2021. It looks like that GHC-9.4 will be second major series GHC release after not bundling `transformers-0.6`.

There also number of issues and pull request been open for year(s), some of which are quite low hanging.

I volunteer myself to become a maintainer of `mtl`. I'd also ask @RyanGlScott if he have time to help too. (Ryan is a maintainer of `transformers-compat` which is closely related).

---

History bit: @ekmett  "donated" mtl to CLC https://github.com/haskell/mtl/issues/69#issuecomment-611418931 and https://github.com/haskell/mtl/issues/69#issuecomment-778763644. Note, there haven't been `mtl` release since that issue was opened (December 2019), when `transformers` got new monad transformers (`AccumT` and `SelectT`). There is an open PR adding corresponding type-classes to `mtl`.

**Comment by ekmett** [comment]
Thanks, Oleg. I don't have an objection in principle. I'd like the assent of the CLC if this were to happen, as I did after all, hand things off to them.

**Comment by tomjaguarpaw** [comment]
From my point of view a summary of the current situation and our possible routes forward would be very helpful regarding making a decision.

**Comment by phadej** [comment]
@tomjaguarpaw I'm not sure I understand what you ask. The current situation is that *mtl is in neglected state*. I propose one route (i.e. *install new maintainers*). I don't know if there are any way to *resolve situation by other means.*. That's up to CLC.

**Comment by ekmett** [comment]
Here's a general summary of the status quo.

`transformers` 0.6 was released. It contained a few new types that had somewhat complicated relationships with the existing abstractions in the `mtl`. So now there becomes the question of whether we should try to capture the new semantics of MonadAccum and MonadSelect, ad/or adapt the existing classes where they can fit. This means teasing out what the right laws are for those new abstractions.

But then there are also some long term niggling issues, like the fact that the `mtl` as it is currently shipped isn't in line with the PVP due to the re-exports of several `base` modules, which makes it so every time base expands `Control.Monad` or the like with any extra methods, every `mtl` user has to either hide the same import a dozen times, or move to qualified imports.

There's also been a longstanding desire to split up some of the `mtl` classes like MonadReader into the "effect" part (like `ask`) and the "handler" parts, (like `local`), as e.g.

    instance Reifies s e => MonadAsk e (Tagged s a)

could allow you to ask but would not provide `local`. Similarly, you can come up with not-quite-writers that can log to a database or file, that do not support `pass` or `listen`.

The plan that the CLC had was that they'd ship a `mtl` 2.3 that mostly acted as a maintenance release, and ripped off the bandaid of the re-exports, then follow up with an `mtl` 3 release that included the new classes, and which finally refactored the basic `mtl` classes to allow these simpler 'effect' type superclasses, e.g. factoring out a `MonadAsk` superclass of `MonadReader`, `MonadTell` superclass of `MonadWriter`, etc.

Those latter bits could then be useful to folks who care about effect systems more broadly than just the `mtl`, which being useful in their own right.

So to facilitate that I handed the `mtl` over to the CLC, right as the CLC decided to reform their internal working flow, processes and membership. Then due to that, and then Emily departing the HF for parts unknown this has languished in the vine for quite a while. I can sympathize with Oleg for wanting someone actively working on this.

My understanding is that at this time there is a 2.3 release candidate pushed to hackage, but not an actual 2.3 release.

**Comment by andreasabel** [comment]
I appreciate the initiative and getting the discussion on `mtl` rolling again.

@ekmett wrote:
> My understanding is that at this time there is a 2.3 release candidate pushed to hackage, but not an actual 2.3 release.

Yes, the latest release candidate RC4 should now simply be released.

It has shown that RC3 which _ripped off the bandaid of the re-exports_ just causes too much breakage to be released without long deprecation periods.  And unfortunately, GHC lacks the facilities to deprecate importing stuff via reexports (rather than importing it from the actual source).

In my perspective, `mtl` is a very basic package only matched by `containers` maybe, and as such it should be handled with __utmost conservativity__.  So it should be on par with `base` in that respect.  It is not a package that just can have some maintainers who make internal decisions.

It's great that there are plans to develop monad transformers further.  But I think this should happen in an _entirely new package_ (working title `mtl-new`) that may and should strive for some compatibility with `mtl`, but require active opting-in from the users (switching their dependency from `mtl` to `mtl-new`).

We know that with stackage's one-version policy just a major-major bump to `mtl-3` does not achieve the same as having a new package `mtl-new`. If you make drastic changes to `mtl`, __users cannot opt out__, and it will cause major disruption.

**Comment by chessai** [comment]
This is something I wanted to bring up after my vacation, but I will do it now, because I don't want to make people wait even longer than I've already made them wait:

I no longer have the time nor energy to devote to MTL, and I suspect neither does Emily, based on our conversations. It's clear that we need a new set of maintainers who have the time and energy.

@phadej and @RyanGlScott were definitely in my list of candidates, I have no problem with them maintaining MTL.

**Comment by Ericson2314** [comment]
> It has shown that RC3 which ripped off the bandaid of the re-exports just causes too much breakage to be released without long deprecation periods. And unfortunately, GHC lacks the facilities to deprecate importing stuff via reexports (rather than importing it from the actual source).

A bit off-topic. I would very much like to see this fixed! In general, besides behind-the-scenes cleanups to GHC, I would like to see the HF focus on this and other such tools to make migrations easier. Then all this good maintainence work is so much less grueling, and we can get more stuff down while wearing down our volunteers less.

**Comment by ekmett** [comment]
Re: Stability 

The `mtl` has been remarkably stable. It was released in 2006. There was a big set of breakages/changes in 2010. Then I've left it alone for _12 years_.

There have been a lot of things we've wanted to do, but deliberately held off because on their own they weren't worth incurring pain for users.

The primary reason why there is interest in shipping a new version at all is that the bedrock upon which it builds (`transformers`) shifted rather drastically 3 years ago bringing all sorts of new things into scope and folks would like to work with those things. That forces some API redesign and a major version bump.

If we're going that far, everyone who uses the library is going to have to at the very least change some things. 

Fixing the re-exports fixes a whole class of constant low-key breakages happening with every few GHC releases, in exchange for a one-time pass.

The reason I've by and large stepped away from maintaining the `mtl` in particular, and to some extent my libraries more broadly, and it has fallen more and more on Ryan and Oleg is, frankly, exactly this. I can do nothing, and have issues accumulate indefinitely with low key discontent, or I can act, and have an actively angry minority shout about how I'm a bad maintainer because I changed something. After all, I threw the lever on the trolley problem, and now it is actively my fault, rather than passively so.

It seems to me that it is important to distinguish between changing something just for the sake of changing it, and changing something to make it better. I do think both of these changes being discussed cross the latter bar. I endeavor to try to accommodate differing opinions, but I think having the ability to change things when warranted is better than permanent ossification.

**Comment by ekmett** [comment]
@Ericson2314 The ability to deprecate re-exports was something that we requested as the CLC very early on. It never happened though, so in the absence of such a feature we can only do what we can, and I really am not in a hurry to hang _any_ plan on a feature request that has already languished for 8 years.

**Comment by chessai** [comment]
> I can do nothing, and have issues accumulate indefinitely with low key discontent, or I can act, and have an actively angry minority shout about how I'm a bad maintainer because I changed something. After all, I threw the lever on the trolley problem, and now it is actively my fault, rather than passively so.

I can relate to this, it's why I've also transitioned away from maintenance of core packages. Even in the cases where there isn't outright anger, there's a lot of loud noise. It feels paralysing, and in some cases, makes you feel alienated/othered. There is a wider problem to discuss there which I have a lot of thoughts about, but this is not the forum for that.

**Comment by tomjaguarpaw** [comment]
> If one long term possibility is that we remove certain re-exports from mtl modules it seems like the documentation should warn about this. Of course, this is not as good as deprecating the re-exports, but that functionality doesn't exist, and in its absence it seems like this would be an important way to communicate future potential breakage to users.

https://github.com/haskell/mtl/pull/106

**Comment by ekmett** [comment]
Fair point. That is a point in the design space, but it does have the disadvantage that the space of what you get in scope becomes harder and harder to articulate over time.

**Comment by phadej** [comment]
I politely ask to discuss how to bring life to `mtl`  package, not what *someone* could/should do there. That *someone* is missing, and that is the problem.

Also `mtl-2.3` is not only about re-exports stuff. Perfectly all of checkboxes in https://github.com/haskell/mtl/issues/86 would be ticked. (That's not much work, but work nevertheless).

**Comment by Bodigrim** [comment]
+1 to @phadej's last comment, let's keep this discussion limited to finding active maintainers for `mtl`. Other questions (such as breaking changes in `mtl`) are better discussed at `mtl` issue tracker itself. 

To summarise, @ekmett as a former maintainer and @chessai as an acting maintainer are in favor of @phadej acquiring maintainership of `mtl`. @emilypi, as another acting maintainer, could you please chime in? If we do not hear from you within a week, I think it would be fair to imply that you do not object.

**Comment by Ericson2314** [comment]
(Sorry to derail any further but) @ekmett I do agree in not wanting to *wait* for tools, just wanting to make those tools finally exist. In the Data.List thread https://github.com/haskell/core-libraries-committee/issues/22#issuecomment-1086701500 I was advocating the basically the same thing --- make the changes that should be made, and then use the new tooling to broker peace with the pitchforks.

**Comment by tomjaguarpaw** [comment]
> let's keep this discussion limited to finding active maintainers for `mtl`

It's not quite a simple as that. I am considering submitting my candidacy for `mtl` maintainer but to decide I need to have a clear understanding of exactly what work is outstanding.

**Comment by Bodigrim** [comment]
AFAICT here are tasks in the order of importance/urgency:
1. Cut a new release compatible with [`MonadTrans`](https://hackage.haskell.org/package/transformers-0.6.0.4/docs/Control-Monad-Trans-Class.html#t:MonadTrans) from `transformers-0.6`.
2. Decide what to do with re-exports from `Data.Monoid` and `Control.Monad`.
3. Support new transformers such as `AccumT` and `SelectT`.
4. Consider viability of reshuffling the hierarchy of type classes with `MonadAsk` and friends.

**Comment by emilypi** [comment]
Oof. @chessai and I completely off the wagon for various reasons over the past few months - apologies folks. I'm happy to pick this up again and get it to a reasonable state now that. 

I can pick this up again and push through any changes needed over the next week. As far as maintenance goes, I'd love to still be on board, but life is happening to me, so an active co-maintainer to whip me would be in order.

**Comment by ekmett** [comment]
@emilypi Concretely, are you okay with accepting Oleg as a co-maintainer here?

**Comment by emilypi** [comment]
Considering Oleg's already massive backlog of maintenance projects, I think it would be irresponsible to add more to his plate without at least attempting to diversify. I would vote @RyanGlScott or a headhunt first before defaulting to Oleg.

**Comment by tomjaguarpaw** [comment]
I'm happy to tackle [Bodigrim's 1 and 2](https://github.com/haskell/core-libraries-committee/issues/52#issuecomment-1090726296), that is

1. Cut a new release compatible with [`MonadTrans`](https://hackage.haskell.org/package/transformers-0.6.0.4/docs/Control-Monad-Trans-Class.html#t:MonadTrans) from `transformers-0.6`.
2. Decide what to do with re-exports from `Data.Monoid` and `Control.Monad`.

[contingent upon these tasks not being more complicated than my initial impression of them, which I will confirm by appropriate due diligence before starting]

As I understand it "Cut a new release compatible with [MonadTrans](https://hackage.haskell.org/package/transformers-0.6.0.4/docs/Control-Monad-Trans-Class.html#t:MonadTrans) from transformers-0.6." means "Cut a new release compatible with the new `MonadTrans t` which has the morally correct quantified superclass constraint `forall m. Monad m => Monad (t m)`". Please correct me if I'm wrong.

**Comment by Kleidukos** [comment]
> > It has shown that RC3 which ripped off the bandaid of the re-exports just causes too much breakage to be released without long deprecation periods. And unfortunately, GHC lacks the facilities to deprecate importing stuff via reexports (rather than importing it from the actual source).
> 
> A bit off-topic. I would very much like to see this fixed! In general, besides behind-the-scenes cleanups to GHC, I would like to see the HF focus on this and other such tools to make migrations easier. Then all this good maintainence work is so much less grueling, and we can get more stuff down while wearing down our volunteers less.

@mpickering Do you think it would be possible to add this to the agenda for 9.6? :)

**Comment by emilypi** [comment]
Sold.

**Comment by Bodigrim** [comment]
Let's not rush. 

1. According to https://hackage.haskell.org/package/mtl/maintainers/ and similar to other Ed's packages, @RyanGlScott is a maintainer of `mtl` already. @RyanGlScott do you have interest/bandwidth to resume your maintainership over `mtl`? Given your experience with `transformers-compat` and a tremendous track record of supporting key packages, I'd be happy to know that `mtl` is back in your hands. Since we are not installing a new maintainer, this option does not require a CLC vote.

2. If Ryan is not available, there are at least three volunteers in this thread (@phadej, @tomjaguarpaw, @kozross). CLC should hold a (ranked) vote to select one of them.

**Comment by Ericson2314** [comment]
I do like a situation where the CLC keeps it's powers in "reserve" / in case any huge disputes need mediation, but there are more active per library appointees in charge of the day-to-day with sufficient devolved powers to not be mere grunts. I think that model would be good for more than just the `mtl`.

**Comment by RyanGlScott** [comment]
Sorry for not chiming in earlier.

I'm no longer on the Core Libraries Committee, so I'm not really a maintainer at the moment. I'm reluctant to volunteer to maintain yet another library, since (1) I already maintain quite a bit, and (2) I think there are other qualified people who would be great at the role (and likely do a better job than I would).

**Comment by Bodigrim** [comment]
> If Ryan and Emily are the current official maintainers, shouldn't they decide whom to add?

Not quite so in general. CLC is responsible to uphold a high level of maintenance, and such responsibility comes with rights necessary to fulfill it, potentially installing a maintainer against wishes of incumbents. We had such precedents in the past. At the very least, CLC must hold a vote of confidence, confirming that a suggested maintainer is sufficient to resolve the issue.

In this spefic case current maintainers are chessai and Emily, but not Ryan. Since both of them are CLC members themselves, it would be unexpected if CLC as a body does not adhere to their opinion, so the paragraph above is mostly a theoretical speculation.

It would be nice to have more than one active maintainer, but not all potential candidates necessarily can work well together. Since we are not revoking rights from the existing team, I think adding one more could be enough for now. 

My personal preference is a ranked vote among volunteers. Alternatively, if @chessai, @emilypi and ideally @ekmett can quickly come up with a common preference, CLC can hold a vote of confidence in this person.

**Comment by Bodigrim** [comment]
@chessai, @emilypi could you please together recommend a new maintainer for `mtl`?

**Comment by emilypi** [comment]
@kozross

**Comment by Bodigrim** [comment]
@chessai could you please confirm that you are also onboard with @kozross and happy to recommend him?

**Comment by chessai** [comment]
> @kozross 

+1

**Comment by ekmett** [comment]
I'll triple the @kozross suggestion for the record.

**Comment by Bodigrim** [comment]
Awesome. Dear CLC members, @tomjaguarpaw, @cgibbard, @cigsender, following the endorsement from current and former maintainers, let's hold a vote of confidence in @kozross as a new (additional) maintainer for `mtl`. @chessai and @emilypi, I'll count your endorsements as votes in favor. 

----

+1 from me.

**Comment by tomjaguarpaw** [comment]
+1 from me

----

My offer to tackle the following two tasks remains, should Koz (if appointed) want to take me up on that.

1. Cut a new release compatible with [MonadTrans](https://hackage.haskell.org/package/transformers-0.6.0.4/docs/Control-Monad-Trans-Class.html#t:MonadTrans) from transformers-0.6.
2.  Decide what to do with re-exports from Data.Monoid and Control.Monad.

**Comment by kozross** [comment]
@tomjaguarpaw Sure, I'm happy for you to do those. I really reckon the correct move is to remove the re-exports: this is well within the scope of a major version bump.

**Comment by cigsender** [comment]
+1 for @kozross -- and if another maintainer is required in the future, I may submit my own name at that juncture :)

**Comment by chessai** [comment]
> Thanks all, @kozross is unanimously approved as a new maintainer of `mtl`. 
> 
> @chessai @emilypi @ekmett could someone of you please grant @kozross necessary rights for GitHub repo and Hackage package?

@kozross you have now been added as a maintained on the haskell/mtl repository.
